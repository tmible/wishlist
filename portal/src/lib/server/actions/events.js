/** @module События действий */

/**
 * Событие добавления действия
 * @constant {string}
 */
export const AddAction = {};
