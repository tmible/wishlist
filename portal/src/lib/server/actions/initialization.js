import { initAddActionStatement } from './statements/add-action.js';

/**
 * Подготовка SQL выражений для работы с действиями
 * @function initActionsFeature
 * @returns {void}
 */
export const initActionsFeature = () => {
  initAddActionStatement();
};
