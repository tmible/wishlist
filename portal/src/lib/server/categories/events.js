/** @module События категорий */

/**
 * Событие получения категорий
 * @constant {string}
 */
export const GetCategories = 'get categories';

/**
 * Событие добавления категории
 * @constant {string}
 */
export const AddCategory = 'add category';

/**
 * Событие обновления категории
 * @constant {string}
 */
export const UpdateCategory = 'update category';

/**
 * Событие удаления категории
 * @constant {string}
 */
export const DeleteCategory = 'delete category';
