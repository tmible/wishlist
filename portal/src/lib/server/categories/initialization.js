import { initAddCategoryStatement } from './statements/add-category.js';
import { initDeleteCategoryStatement } from './statements/delete-category.js';
import { initGetCategoriesStatement } from './statements/get-categories.js';
import { initUpdateCategoryStatement } from './statements/update-category.js';

/** @typedef {{ id: number, name: string }} Category */

/**
 * Подготовка SQL выражений для работы с категориями
 * @function initCategoriesFeature
 * @returns {void}
 */
export const initCategoriesFeature = () => {
  initGetCategoriesStatement();
  initAddCategoryStatement();
  initUpdateCategoryStatement();
  initDeleteCategoryStatement();
};
