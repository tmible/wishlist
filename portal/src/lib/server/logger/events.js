/** @module События логгера */

/**
 * Событие создания промежуточного обработчика, логирующего запросы и ответы
 * @constant {string}
 */
export const CreateLoggingMiddleware = 'create logging middleware';
