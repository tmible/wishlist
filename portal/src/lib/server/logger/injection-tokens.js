/** @module Токены внедрения адаптера логирования */

/**
 * Токен внедрения логгера
 * @constant {string}
 */
export const Logger = 'logger';
