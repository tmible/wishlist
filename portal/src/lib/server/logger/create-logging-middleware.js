import { inject } from '@tmible/wishlist-common/dependency-injector';
import {
  UNKNOWN_USER_UUID_COOKIE_NAME,
} from '$lib/constants/unknown-user-uuid-cookie-name.const.js';
import { Logger } from './injection-tokens.js';

/** @typedef {import('@sveltejs/kit').RequestEvent} RequestEvent */
/**
 * @template T
 * @typedef {import('@sveltejs/kit').MaybePromise<T>} MaybePromise
 * @see {@link https://svelte.dev/docs/kit/@sveltejs-kit#MaybePromise}
 */

/** @module Создание промежуточного обработчика, логирующего запросы и ответы на них */

/**
 * @typedef {Function} LoggingMiddleware
 * @param {RequestEvent} event Событие запроса
 * @param {() => MaybePromise<Response>} next Функция вызова следующих обработчиков
 * @returns {Promise<Response>} Ответ от следующих обработчиков
 * @async
 */

/**
 * Создание промежуточного обработчика
 * @function createLoggingMiddleware
 * @returns {LoggingMiddleware} Промежуточный обработчик
 */
export const createLoggingMiddleware = () => {
  const logger = inject(Logger);
  return async (event, next) => {
    const requestUuid = crypto.randomUUID();
    event.locals.requestUuid = requestUuid;

    logger.info(
      { requestUuid, unknownUserUuid: event.cookies.get(UNKNOWN_USER_UUID_COOKIE_NAME) ?? null },
      `request ${
        event.request.method
      } ${
        event.url.pathname
      }; cookie: ${
        event.cookies.getAll().map(({ name, value }) => `${name}=${value}`).join(';')
      }; body: ${
        event.request.body === null ? null : await event.request.clone().text()
      }`,
    );

    const response = await next(event);

    let responseBody = null;
    if (response.body !== null) {
      responseBody =
        response.body instanceof ReadableStream ?
          'event-stream' :
          (await response.clone().text());
    }

    logger.info(
      {
        requestUuid,
        unknownUserUuid: event.cookies.get(UNKNOWN_USER_UUID_COOKIE_NAME) ?? null,
        userid: event.locals.userid ?? null,
      },
      `response ${response.status}; body: ${responseBody}`,
    );

    return response;
  };
};
