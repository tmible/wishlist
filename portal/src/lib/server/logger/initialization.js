import { resolve } from 'node:path';
import { provide } from '@tmible/wishlist-common/dependency-injector';
import { subscribe } from '@tmible/wishlist-common/event-bus';
import pino from 'pino';
import { LOGS_DB_FILE_PATH, LOGS_DB_MIGRATIONS_PATH } from '$env/static/private';
import { createLoggingMiddleware } from './create-logging-middleware.js';
import { CreateLoggingMiddleware } from './events.js';
import { Logger } from './injection-tokens.js';

/**
 * Инициализация логгера. Логи с уровнем INFO и выше пишутся в БД.
 * Логи с уровнем DEBUG и выше пишутся в стандартный вывод.
 * Регистрация логгера в сервисе внедрения зависимостей.
 * Подписка на событие создания промежуточного обработчика
 * @function initLogger
 * @returns {void}
 */
export const initLogger = () => {
  const logger = pino(
    { level: 'debug' },
    pino.transport({
      targets: [
        { target: resolve(import.meta.dirname, 'pino-sqlite-transport.worker.js'), level: 'info' },
        { target: 'pino-pretty', level: 'debug', options: { destination: 1 } },
      ],
      worker: {
        env: {
          LOGS_DB_FILE_PATH,
          LOGS_DB_MIGRATIONS_PATH,
        },
      },
    }),
  );

  provide(Logger, logger);
  subscribe(CreateLoggingMiddleware, createLoggingMiddleware);
};
