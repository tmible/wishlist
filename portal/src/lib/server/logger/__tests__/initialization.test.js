import { dirname } from 'node:path';
import { provide } from '@tmible/wishlist-common/dependency-injector';
import { subscribe } from '@tmible/wishlist-common/event-bus';
import pino from 'pino';
import { afterEach, describe, expect, it, vi } from 'vitest';
import { createLoggingMiddleware } from '../create-logging-middleware.js';
import { CreateLoggingMiddleware } from '../events.js';
import { initLogger } from '../initialization.js';
import { Logger } from '../injection-tokens.js';

vi.mock('@tmible/wishlist-common/dependency-injector');
vi.mock('@tmible/wishlist-common/event-bus');
vi.mock('pino');
vi.mock(
  '$env/static/private',
  () => ({
    LOGS_DB_FILE_PATH: 'LOGS_DB_FILE_PATH',
    LOGS_DB_MIGRATIONS_PATH: 'LOGS_DB_MIGRATIONS_PATH',
  }),
);

describe('logger / initialization', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should initalize logger', () => {
    vi.mocked(pino).transport = vi.fn((...args) => [ 'pino.transport', ...args ]);

    initLogger();

    // Замена вручную, чтобы тест проходил, несмотря на разные пути в разных средах
    const [ target ] = vi.mocked(pino).mock.lastCall[1][1].targets;
    target.target = target.target.replace(dirname(import.meta.dirname), '/import/meta/dirname');

    expect(vi.mocked(pino).mock.lastCall).toMatchSnapshot();
  });

  it('should provide logger', () => {
    vi.mocked(pino).mockReturnValueOnce('logger');
    initLogger();
    expect(vi.mocked(provide)).toHaveBeenCalledWith(Logger, 'logger');
  });

  it('should subscribe to CreateLoggingMiddleware event', () => {
    initLogger();
    expect(
      vi.mocked(subscribe),
    ).toHaveBeenCalledWith(
      CreateLoggingMiddleware,
      createLoggingMiddleware,
    );
  });
});
