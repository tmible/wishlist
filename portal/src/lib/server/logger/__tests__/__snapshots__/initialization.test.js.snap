// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`logger / initialization > should initalize logger 1`] = `
[
  {
    "level": "debug",
  },
  [
    "pino.transport",
    {
      "targets": [
        {
          "level": "info",
          "target": "/import/meta/dirname/pino-sqlite-transport.worker.js",
        },
        {
          "level": "debug",
          "options": {
            "destination": 1,
          },
          "target": "pino-pretty",
        },
      ],
      "worker": {
        "env": {
          "LOGS_DB_FILE_PATH": "LOGS_DB_FILE_PATH",
          "LOGS_DB_MIGRATIONS_PATH": "LOGS_DB_MIGRATIONS_PATH",
        },
      },
    },
  ],
]
`;
